src/PostgREST/DbStructure.hs

```diff
@@ -738,36 +731,100 @@ pfkSourceColumns cols pgVer =
+            -- to be able to use regexp_replace() only once.
+            -- This has been tested against a huge schema with 250+ different views.
+            -- The unit tests do NOT reflect all possible inputs. Be careful when changing this!
+            -- -----------------------------------------------
+            -- pattern           | replacement         | flags
+            -- -----------------------------------------------
+            -- `,` is not part of the pg_node_tree format, but used in the regex.
+            -- This removes all `,` that might be part of column names.
+               ','               , ''
+            -- The same applies for `(`, `{` and `}`, although those are used a lot in pg_node_tree.
+            -- We remove the escaped ones, which might be part of column names again.
+            ), '\('              , ''
+            ), '\{'              , ''
+            ), '\}'              , ''
+            -- The fields we need are formatted as json manually to protect them from the regex.
+            ), ' :targetList '   , ',"targetList":'
+            ), ' :resno '        , ',"resno":'
+            ), ' :resorigtbl '   , ',"resorigtbl":'
+            ), ' :resorigcol '   , ',"resorigcol":'
+            -- Make the regex also match the node type, e.g. `{QUERY ...`, to remove it in one pass.
+            ), '{'               , '{ :'
+            -- Protect node lists, which start with `({` or `((` from the greedy regex.
+            -- The extra `{` is removed again later.
+            ), '(('              , '{(('
+            ), '({'              , '{({'
+            -- This regex removes all unused fields to avoid the need to format all of them correctly.
+            -- This leads to a smaller json result as well.
+            -- Removal stops at `,` for used fields (see above) and `}` for the end of the current node.
+            -- Nesting can't be parsed correctly with a regex, so we stop at `{` as well and
+            -- add an empty key for the followig node.
+            ), ' :[^{,}]+'       , ',"":'              , 'g'
+            -- For performance, the regex also added those empty keys when hitting a `,` or `}`.
+            -- Those are removed next.
+            ), ',"":}'           , '}'
+            ), ',"":,'           , ','
+            -- This reverses the "node list protection" from above.
+            ), '{('              , '('
+            -- Every key above has been added with a `,` so far. The first key in an object doesn't need it.
+            ), '{,'              , '{'
+            -- pg_node_tree has `()` around lists, but JSON uses `[]`
+            ), '('               , '['
+            ), ')'               , ']'
+            -- pg_node_tree has ` ` between list items, but JSON uses `,`
+            ), '} {'             , '},{'
+            ), '] ['             , '],['
+            -- `<>` in pg_node_tree is the same as `null` in JSON, but due to very poor performance of json_typeof
```

Comment:
> but due to very poor performance of json_typeof

@wolfgangwalther Oh, how bad is it? We're using json_typeof for our INSERTs/UPDATEs.

Reply:
I remember this was really bad. I might be able to do some testing, to confirm that. With `json_typeof` the query looked a bit different - so maybe it wasn't directly json_typeof, but the other query structure in general.

Reply:
So I just tested this for this query again. The changes I made where:
```
), '<>'              , '[]'
```
becomes
```
), '<>'              , 'null'
```
(so no target list is encoded as null instead of an empty array)

and then

```
target_entries as(
        select
          view_id, view_schema, view_name,
          json_array_elements(view_definition->0->'targetList') as entry
        from transform_json
        where json_typeof(view_definition->0->'targetList') = 'array' -- this line is added
      ),
```

so empty target lists are now filtered by json_typeof.

The changed replace should have no effect, so only json_typeof makes things slower here. Here are the results:

| query | execution |
| -- | --: |
| old master | 415.0 |
| master | 232.6 |
| json PR | 321.9 |
| json_typeof | 714.1 |

So just the json_typeof call takes more time then everything else combined - the time is more than doubled!

Reply:
I tested a bit more and even without `json_typeof`, just adding the following make the query just as slow:

```
where view_definition::text is not NULL
```

This is effectively a no-op - but the cast to `text` is taking quite a bit of time. It probably has to go through the whole json object. The same thing will happen here for `json_typeof`.

We're currently using json_typeof here:
https://github.com/PostgREST/postgrest/blob/11d62a80105c21696e65b70acf07b6f47405ca68/src/PostgREST/Private/QueryFragment.hs#L40-L53

Maybe we can just check the first non-whitespace character of the body we're passing in, in haskell. If it's a `[` use the array code, if it's a `{` use the object code. Could give us a performance improvement. Or not. Would have to loadtest this, of course :)

Reply:
Ahh.. I see it now! Maybe we could make the `[ or {` check  on pg?

I feel like messing with the payload on Haskell might bring https://github.com/PostgREST/postgrest/issues/690 again.

Or perhaps we should just accept the `json_typeof` overhead. It's nice to be aware of it though.

Reply:
> Ahh.. I see it now! Maybe we could make the `[ or {` check on pg?

Yeah, I thought so. That was the reason why I tried `where view_definition::text is not NULL`. Because to do the check, we need to cast the json back to text first. Just as slow as `pg_typeof`.

> I feel like messing with the payload on Haskell might bring #690 again.

Yeah, we would need to make sure that we're not parsing the body at all. We just want to read the first bytes of it. Isn't that kind of lazy evaluation one of haskell's strength? If we do it properly of course.

That still does not mean the json is correct in any way. But in that case PG would still throw.

Plus, it would also solve #1323. I tried to debug that today, but I don't see why this is happening. This would be fixed by:
- if `[` -> recordset
- if `{` -> record
- else -> error

Reply:
Still had this one my todo list - but I guess we can call this resolved. My tests were flawed, it wasn't really `json_typeof` that was performing that bad, it was the "accessing nested json values" part that came with the approach.

`json_typeof` does have a little bit of overhead - but that's to be expected when it's called ~2800 times when querying the schema cache in *the big schema*. I don't see a problem when it's called once per request.